src/Raven.Server/Documents/RetireAttachmentsSender.cs

```diff
@@ -115,32 +107,66 @@ internal async Task<int> RetireAttachments(int batchSize, long maxItemsToProcess
                             return totalCount;
                         }
 
-                        using (directUpload.Initialize())
+                        var directUploaders = new Dictionary<string, DirectFileUploader>();
+
+                        try
                         {
                             // upload the attachments to cloud and update the document
                             foreach (var doc in toRetire)
                             {
                                 _token.ThrowIfCancellationRequested();
 
-                                var key = doc.LowerId.ToString();
-                                var collection = doc.Id;
-
                                 if (CanContinueBatch(Logger, duration, totalUploaded) == false)
                                 {
                                     break;
                                 }
 
-                                if (string.IsNullOrEmpty(collection))
+                                if (string.IsNullOrEmpty(doc.Id))
                                 {
                                     if (Logger.IsInfoEnabled)
-                                        Logger.Info($"Skipping '{nameof(AttachmentRetireType.PutRetire)}' of retired attachment with key: '{key}' because it's collection '{collection}' IsNullOrEmpty.");
+                                        Logger.Info($"Skipping PutRetire of retired attachment with key: '{doc.LowerId}' because it's identifier '{doc.Id}' IsNullOrEmpty.");
 
                                     // document was deleted, need to remove it from retired tree
                                     retired.Enqueue(doc);
                                     continue;
                                 }
 
-                                var hash = _database.DocumentsStorage.AttachmentsStorage.ExtractHashFromAttachmentId(doc.LowerId);
+
+                                // get uploader for the attachment
+                                if (directUploaders.TryGetValue(doc.Id, out var directUpload) == false)
+                                {
+                                    // check if we have a destination configuration for the identifier
+
+                                    if (Configuration.Destinations.TryGetValue(doc.Id, out var destination) == false)
+                                    {
+                                        // no destination found for the identifier
+                                        if (Logger.IsWarnEnabled)
+                                            Logger.Warn($"No destination found for retired attachment with key: '{doc.LowerId}' and identifier '{doc.Id}'. Will try to retire it again!");
```

Comment:
when will this happen? 
What if I have 1M such attachments, and no destination? Will it have to scan them each time?